Software/Program/Code/Core/Src/main.c

```diff
@@ -40,24 +41,80 @@
 /* USER CODE END PM */
 
 /* Private variables ---------------------------------------------------------*/
+ADC_HandleTypeDef hadc3;
 
 TIM_HandleTypeDef htim2;
 
 /* USER CODE BEGIN PV */
 
+/* Public variables ---------------------------------------------------------*/
+static volatile bool OnHighPulse;
+static volatile bool ButtonPressed;
+static volatile bool ADC_ConversionComplete;
+static uint32_t ARR_Reload_Values[] = { ARR_TABLE_INITIALIZATION_VALUE, ARR_TABLE_INITIALIZATION_VALUE };
+static uint8_t ARR_Reload_Value_Idx;
+
 /* USER CODE END PV */
 
 /* Private function prototypes -----------------------------------------------*/
 void SystemClock_Config(void);
 static void MX_GPIO_Init(void);
 static void MX_TIM2_Init(void);
+static void MX_ADC3_Init(void);
 /* USER CODE BEGIN PFP */
 
 /* USER CODE END PFP */
 
 /* Private user code ---------------------------------------------------------*/
 /* USER CODE BEGIN 0 */
 
+void HAL_TIM_PeriodElapsedCallback( TIM_HandleTypeDef * timer_handle )
+{
+   if ( timer_handle == &htim2 )
+   {
+      HAL_GPIO_TogglePin( FRQ_OUT_PORT, FRQ_OUT_PIN );
+      
+      // Load next ARR value into ARR
+      __HAL_TIM_SET_AUTORELOAD( &htim2, ARR_Reload_Values[ ARR_Reload_Value_Idx ] );
+      
+      // Update index for next run
+      ARR_Reload_Value_Idx = ( ARR_Reload_Value_Idx + 1 ) % (sizeof(ARR_Reload_Values) / sizeof(uint32_t));
+   }
+}
+
+/**
+  * @brief  EXTI line detection callback.
+  * @param  GPIO_Pin: Specifies the port pin connected to corresponding EXTI line.
+  * @retval None
+  */
+void HAL_GPIO_EXTI_Callback(uint16_t GPIO_Pin)
+{
+   if ( GPIO_Pin == B1_Pin )
+   {
+      ButtonPressed = true;
+   }
+}
+
+//void HAL_ADC_ConvCpltCallback( ADC_HandleTypeDef * adc_handle )
+//{
+////   // Check if interrupt came from an end-of-conversion (EOC) event...
+////   if( __HAL_ADC_GET_FLAG(adc_handle, ADC_FLAG_EOC) )
+////   {
+////      /* NOTE-WORTHY ADC HAL API:
+////       *    __HAL_ADC_CALC_VREFANALOG_VOLTAGE
+////       *    __HAL_ADC_CALC_DATA_TO_VOLTAGE
+////       *    
+////       */
+////      ADC_ConversionComplete = true;
+////      // No need to clear any ADC registers because that is handled in IRQ routine that calls this callback.
+////   }
+//   ADC_ConversionComplete = true;
+//}
+
+
+/*** Private Helper Functions ***/
+STATIC Hz_T Interpolate_ADCCount_To_Frq( uint16_t adc_count );
```

Comment:
```suggestion
STATIC Hz_T Interpolate_ADCCount_To_Frq( uint16_t adc_count );
```
I'd prefer this to be a look-up table that a downstream user can configure. This will make the logic of this component decouple-able from the `main` function/file.